### PR TITLE
Slack extension (list users, group chats, and channels)

### PR DESCRIPTION
**Pull request by momme-rtf**
## Description

I created an extension that lists all your users, group chats, and channels of your slack workspace and lets you easily open them in your native slack app. In addition to that, the extension will also allow you to access unread messages, snooze notifications, and set your presence status.

## Screencast

slack-01
slack-02
slack-03

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed in the `media` folder

### CODE REVIEW

**Comment by momme-rtf** [comment]
I saw the Github action failed: `Invalid author "momme.rtf". Use the Raycast username`
I'm actually using the same username for Github and also raycast. Maybe you can help me out to fix that issue.

**Comment by momme-rtf** [comment]
@aesign glad to hear that and thanks for the inspiration, might be smth that I would consider for the next iteration of this extension!

**Comment by aesign** [comment]
> @aesign glad to hear that and thanks for the inspiration, might be smth that I would consider for the next iteration of this extension!

You're welcome! Actually, I already got a working POC. For now, i made some edits to the New Message command: I use List's onSelectionChange prop to fetch bookmarks and I use an ActionPanel.Section to display them.

**Comment by thomaspaulmann** [comment]
This is really cool! I built something similar to search users, channels and groups. You might want to take a look at my branch. I use SWR for fetching, which has some nice caching. It also uses hooks which are easy to reason about. Also, I added icons for channels and groups that you might want to use for your extension. Let me know if you need some help with it.

Nit: "New Message" is a slightly confusing name for the command. Isn't it rather a search? You could have an action that actually opens a form to send a message from within Raycast.

**Comment by momme-rtf** [comment]
> You're welcome! Actually, I already got a working POC. For now, i made some edits to the New Message command: I use List's onSelectionChange prop to fetch bookmarks and I use an ActionPanel.Section to display them.

@aesign  Awesome, looking forward to your pr/contribution then! I think the only thing that you should be careful about is the number of API calls when navigating through the list - maybe some debounce mechanism + caching could be helpful.

**Comment by momme-rtf** [comment]
> This is really cool! I built something similar to search users, channels and groups. You might want to take a look at my [branch](https://github.com/raycast/extensions/tree/slack-search).

@thomaspaulmann Cool, will do it!

> I use SWR for fetching, which has some nice caching.

I also implemented caching by myself based on the LocalStorage provided by the raycast-api package. (It reduces the number of queries to one per 6 hours.) 

> Nit: "New Message" is a slightly confusing name for the command. Isn't it rather a search? You could have an action that actually opens a form to send a message from within Raycast.

Hmm, valid point. I followed the naming of Slack itself (when you open `File -> New message`). But I agree that smth that I could refactor.

**Comment by thomaspaulmann** [comment]
> I also implemented caching by myself based on the LocalStorage provided by the raycast-api package. (It reduces the number of queries to one per 6 hours.)

The nice thing about SWR is that it revalidates the cache all the time in the background. Giving the best experience to have everything snappy and up-to-date. It's pretty easy to use and you can check out my code. I also use it in my [Sentry extension](https://www.raycast.com/thomas/sentry).

**Comment by momme-rtf** [comment]
@thomaspaulmann I checked out the implementation examples with `swr`. You're right it's a very interesting package but what I didn't like about the examples is the way how you're accessing the local storage. I would rather want to use the official `LocalStorage` implementation of the `raycast-api` package which also states + ensures that its scope is only for one extension.
I tried to combine `swr` and raycast's `LocalStorage`  but it didn't work out because all the functions of that `LocalStorage` are async...
How do you see using my self-made caching service (which is actually based on raycast's `LocalStorage` ) in that case anyway?

**Comment by thomaspaulmann** [comment]
> @thomaspaulmann I checked out the implementation examples with `swr`. You're right it's a very interesting package but what I didn't like about the examples is the way how you're accessing the local storage. I would rather want to use the official `LocalStorage` implementation of the `raycast-api` package which also states + ensures that its scope is only for one extension.
> I tried to combine `swr` and raycast's `LocalStorage` but it didn't work out because all the functions of that `LocalStorage` are async...
> How do you see using my self-made caching service (which is actually based on raycast's `LocalStorage` ) in that case anyway?

Yeah it's a slightly different way of handling the cache. The cache is stored in the application's support directory, so still scoped to your extension only. The only benefit of the `LocalStorage` would be to have it encrypted in our database. 

I can take a look on how to make this work. I run into similar issues, that's why I simplified it with just file-based caching.

**Comment by momme-rtf** [comment]
> Yeah it's a slightly different way of handling the cache. The cache is stored in the application's support directory, so still scoped to your extension only. The only benefit of the LocalStorage would be to have it encrypted in our database.\n\nI can take a look on how to make this work. I run into similar issues, that's why I simplified it with just file-based caching

Ah, ok. I don't mind - you can also look into it. Maybe you could also squeeze in adding your icons for groups and channels as you suggested;) - I haven't had time to do so today.

File: extensions/slack/src/shared/client/SlackClient.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  }
+  if (a > b) {
+    return 1;
+  }
+  return 0;
+};
+
+export class SlackClient {
+  public static async getUsers(): Promise<User[]> {
+    const users = await LocalStorageService.get<User[]>("users", async (): Promise<User[]> => {
```
Comment (eddyw) [inline extensions/slack/src/shared/client/SlackClient.ts:36]: What if accessToken is updated in preferences and the new token happens to be from a different Slack workspace (users / channels are different)?
I think caching is okay but they should be prefixed (somehow) in a way that changing access token doesn't display cached data from a different Slack workspace

File: extensions/slack/src/shared/client/SlackClient.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      const publicChannels = await slackWebClient.conversations.list({
+        exclude_archived: true,
+        types: "public_channel",
+        limit: 1000,
+      });
+      const privateChannels = await slackWebClient.conversations.list({
+        exclude_archived: true,
+        types: "private_channel",
+        limit: 1000,
+      });
```
Comment (eddyw) [inline extensions/slack/src/shared/client/SlackClient.ts:79]: `conversations.list` accepts `types` as comma separated list of types ([see conversations.list#arg_types](https://api.slack.com/methods/conversations.list#arg_types)), so there is no need to make two requests:
```suggestion
      const channels = await slackWebClient.conversations.list({
        exclude_archived: true,
        types: "public_channel, private_channel",
        limit: 1000,
      });
```
What's more, you can provide all of these:
```
types: "public_channel, private_channel, mpim"
```
So you could have one single request to get private/public channels and groups. Then you can filter whichever needs to be displayed.

File: extensions/slack/src/shared/client/SlackClient.ts (hunk 1 above)
Comment (momme-rtf) [reply extensions/slack/src/shared/client/SlackClient.ts]: Should be fixed with my new changes since I'm now using the package `swr` for caching.

**Comment by momme-rtf** [comment]
@thomaspaulmann I found some time to implement `swr` caching on my own. I also added your icons for groups and channels.

**Comment by thomaspaulmann** [comment]
> @thomaspaulmann I found some time to implement `swr` caching on my own. I also added your icons for groups and channels.

Nice one, this looks really cool! Can't wait to search and jump into Slack quickly.

**Comment by momme-rtf** [comment]
@pernielsentikaer @mattisssa I'm done now with my final tweaks, so feel free to review it.

File: extensions/slack/src/search.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+function SlackList() {
+  const { data: users } = useUsers();
+  const { data: channels } = useChannels();
+  const { data: groups } = useGroups();
+
+  return (
+    <List isLoading={!users || !groups || !channels}>
+      <List.Section title="Users">
+        {users?.map(({ name, id, teamId, icon }) => (
```
Comment (pernielsentikaer) [inline extensions/slack/src/search.tsx:22]:
```suggestion
        {users?.map(({ name, username, id, teamId, icon }) => (
```

Comment (pernielsentikaer) [reply extensions/slack/src/search.tsx:22]: Only a suggestion

File: extensions/slack/src/search.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+          <List.Item
+            key={id}
+            title={name}
```
Comment (pernielsentikaer) [inline extensions/slack/src/search.tsx:25]:
```suggestion
            title={username}
            subtitle={name}
```
Only a suggestion

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @momme-rtf 👋 

Thanks for your contribution, I can't wait to use it myself 💪 

I have some feedback for you to address:

- There is no gracefully handling in `Search` if token is wrong, it just never stops loading and this is what the console says:

```
21:28:41.213 Failed reading cache Error: ENOENT: no such file or directory, open '/Users/pnt/Library/Application Support/com.raycast.macos/extensions/slack/swr-cache'
    at Object.openSync (node:fs:585:3)
    at readFileSync (node:fs:453:35)
    at null.provider (/Users/pnt/Documents/extensions/extensions/slack/src/shared/client/CacheProvider.tsx:14:19)
    at null.<anonymous> (/Users/pnt/Documents/extensions/extensions/slack/node_modules/swr/dist/index.mjs:500:25)
    at mountState (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.development.js:6052:24)
    at Object.useState (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.development.js:6537:20)
    at useState (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react/react.development.js:1533:23)
    at null.SWRConfig$1 (/Users/pnt/Documents/extensions/extensions/slack/node_modules/swr/dist/index.mjs:498:24)
    at renderWithHooks (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.development.js:5637:22)
    at mountIndeterminateComponent (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.development.js:7757:17) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/Users/pnt/Library/Application Support/com.raycast.macos/extensions/slack/swr-cache'
}
    at null.SWRConfig$1 (/Users/pnt/Documents/extensions/extensions/slack/node_modules/swr/dist/index.mjs:492:23)
    at null.CacheProvide<…>
```

- The commands `Set Snooze` and `Set Presence` also dies pretty hard, the right way would be to return an empty array (So it shows “No results” and a failure toast.

zQWpufDgpW6ZJOax

Looking forward to testing it out again 🔥 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

File: extensions/slack/src/search.tsx (hunk 3 above)
Comment (momme-rtf) [reply extensions/slack/src/search.tsx:26]: Hmm, I would prefer to use the name as it is right now

File: extensions/slack/src/search.tsx (hunk 4 above)
Comment (momme-rtf) [reply extensions/slack/src/search.tsx:29]: See answer above

**Comment by momme-rtf** [comment]
Hi @pernielsentikaer
thanks for your review and your thoughts! I implemented the error handling. So you can have a look again. Looking forward to finally seeing it in the store!

File: extensions/slack/src/search.tsx (hunk 3 above)
Comment (pernielsentikaer) [reply extensions/slack/src/search.tsx:26]: Could still make sense to have a subtitle with the username then, maybe it's just me :)

**Review by pernielsentikaer — APPROVED** [review]
Hi @momme-rtf 👋 

Thanks for your contribution 🔥

I'm going to approve it and sending it to the final review before merge

File: extensions/slack/package.json [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    {
+      "name": "search",
+      "title": "Search",
+      "subtitle": "Slack",
+      "description": "Search for all workspace members, group chats, and channels and open them directly in your Slack app.",
+      "mode": "view"
+    },
+    {
+      "name": "unread-messages",
+      "title": "Unread Messages",
```
Comment (mattisssa) [inline extensions/slack/package.json:23]:
```suggestion
      "title": "Open Unread Messages",
```

**Review by mattisssa — CHANGES_REQUESTED** [review]
Hey! Thanks for the extension, reminds me that I must finish my Slack – Unread Messages extension one day... 

One thing I notice is that for some channels/DMs there are no icons. Probably worth using [the fallback images ](https://developers.raycast.com/api-reference/user-interface/icons-and-images#image.fallback) for this case and providing `Icon.Person` for example.

I've also left comment suggestions about one command that I think better to be renamed 😉 





CleanShot 2022-03-25 at 14 59 51@2x

**Review by mattisssa — APPROVED** [review]

**Pull request merged**